Add docs for various source types (#424)
flatpak/flatpak-docs: docs/module-sources.rst

guide will go in depth on each of them and provide examples.

Sources are copied or downloaded to the source directory before the build
starts.

The ``type`` property of a source determines how that source will be
handled before or during build by ``flatpak`` or ``flatpak-builder``.

Most of the sources support ``path`` and ``url`` property where ``path``
is relative to the root of manifest and ``url`` is a direct url to the
source.

The ``url`` property requires a ``checksum`` of the file to be specified
after it using ``sha256`` or ``sha512``.

There are a few properties common to all source types ``only-arches``,
``skip-arches`` and ``dest``.

``only-arches`` and ``skip-arches`` are used to build the source only
on the specified architecture or skip building it on the specified
architecture respectively. This is often useful for binary packages:

.. code-block:: yaml

  modules:
    - name: foo
      buildsystem: simple
      build-commands:
        - ...
        - ...
      sources:
        - type: file
          url: https://example.org/foo-1.0-x86_64.deb
          sha256: 26c9007c655483438d83f17070190fc4b6bccb8ff5efc30fcedbb7d162681a7f
          dest-filename: foo.deb

>           only-arches: [x86_64]

Reviewer: please please, stop advocating mixed syntax.
```suggestion
          only-arches: 
            - x86_64
```

same below.

          sha256: 26c9007c655483438d83f17070190fc4b6bccb8ff5efc30fcedbb7d162681a7f
          dest-filename: foo.tar.xz
          strip-components: 2

- ``mirror-urls`` is a list of alternative urls that will be used for
  downloading, if the main ``url`` fails.

- The ``sha256`` is the sha256 checksum of the downloaded file, supplied
  by upstream of the source or calculated using ``sha256sum``.

- ``dest-filename`` is not usually used for this source. It is used to
  rename the downloaded file. This can be useful if the buildsystem
  expects a certain name for the archive.

- ``strip-components`` is the number of pathname components to strip
  during extraction. This defaults to ``1`` if not specified.

.. code-block::

  # orginal archive
  .
  └── a
      ├── b
      │   ├── c
      │   │   └── FILE
      │   └── FILE
      └── FILE

  # strip-components: 1
  .
  ├── b
  │   ├── c
  │   │   └── FILE
  │   └── FILE
  ├── FILE

  # strip-components: 2
  .
  ├── c
  │   └── FILE
  ├── FILE

``flatpak-builder`` supports
the following archive types: ``"rpm", "tar", "tar-gzip", "tar-compress", "tar-bzip2", "tar-lzip", "tar-lzma", "tar-lzop", "tar-xz", "tar-zst", "zip", "7z"``

> They type of the archive is `gussed` from the suffix of the file basename.

Reviewer: ```suggestion
The archive type is guessed from the suffix of the file basename.
```

``rpm`` is extracted with ``rpm2cpio``, ``zip`` is extracted with ``unzip``
and ``7z`` is extratced with ``7z``. The rest are extracted with

> ``tar``.

Reviewer: It should be mentioned the utilities should be installed on the host for flatpak-builder to be able to use them. 7z usually should be discouraged.

       - ln -s ${FLATPAK_DEST}/bar/executable ${FLATPAK_DEST}/bin/executable
      sources:
        - type: file
          dest-filename: squashed.snap
          url: https://example.org/squashed-1.0.snap
          sha256: <sha256 checksum of squashed-1.0.snap>

``dest-filename`` is typically used with this source type to make the
filename `predictable` for the commands in the ``build-commands`` array.

Patch sources
-------------

These are sources of ``type: patch`` used with patchfiles. The typical
usage is to create patchfiles and place them somewhere relative to the
manifest root. The ``path`` property is the location of the file relative
to the manifest.

.. code-block:: yaml

  modules:
    - name: foo
      buildsystem: cmake-ninja
      builddir: true
      config-opts:
      - ...
    sources:
      - type: git
        url: https://example.org/foo.git
        tag: v1.0.0
        commit: 63c01f14391aef7d7691c7c63a610d47512147ed
      - type: patch
        path: patches/test-patch.patch

Multiple patches can be specified with ``paths``:

.. code-block:: yaml

  - type: patch
    paths:
      - patches/test1-patch.patch
      - patches/test2-patch.patch

> By default, ``patch -p 1`` is used, but it can be adjusted with

Reviewer: We should mention that the default are suited for the default format that git generates.

``strip-components`` and additional options can be passed with
``options`` :

.. code-block:: yaml

  - type: patch
    paths:
      - patches/test1-patch.patch
      - patches/test2-patch.patch
    options:
      - -d
      - dir1

``use-git: true`` and ``use-git-am: true`` can be used to use
``git apply``and ``git am`` respectively instead of ``patch``.

Shell and Script sources
------------------------

A shell source with ``type: shell`` is used to modify a source during
extraction. ``commands`` takes in any shell commands.

This is usually used to make quick modifications to sources but a patch
should be used for any complex modification to make it fail safe.

.. code-block:: yaml

  modules:
    - name: foo
      sources:
        - type: archive
          url: https://example.org/foo.tar.xz
          sha256: ea029a2e21d2d6ad0a156f6679bd66836204aa78148a4c5e498fe682e77127ef
        - type: shell
          commands:

>             - sed -i 's|/usr/local|/app|g' makefile.common

Reviewer: I'd rather use the example of the autoconf files. Using `sed` should be discouraged in general.

- ``size`` should be the file size of the package in `bytes` (``wc --bytes < file``).

- ``filename`` can be anything to name the downloaded source package.

The binary source packages are often specific to one architecture, so
``only-arches: [x86_64]`` or ``only-arches: [aarch64]`` can be used to
make the build specific to that architecture only.

A typical extra-data manifest may look like:

.. code-block:: yaml

  modules:
    - shared-modules/lzo/lzo.json
    - shared-modules/squashfs-tools/squashfs-tools.json

    - name: dependency1
      ...

    - name: dependency2
      ...

    - name: foo
      buildsystem: simple
      build-commands:
        - install -Dm644 ${FLATPAK_ID}.metainfo.xml ${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.metainfo.xml
        - install -Dm644 foo.desktop ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
        - install -Dm644 foo.svg ${FLATPAK_DEST}/share/icons/hicolor/scalable/apps/${FLATPAK_ID}.svg
        - ln -s ${FLATPAK_DEST}/extra/subdir/foo_binary ${FLATPAK_DEST}/bin/foo_binary
        - install -Dm755 apply_extra ${FLATPAK_DEST}/bin/apply_extra
      sources:
        - type: extra-data
          filename: foo.snap

>           only-arches: [x86_64]

Reviewer: ```suggestion
          only-arches: 
            - x86_64
```

            - rm -r squashfs-root foo.snap

The ``buildsystem`` is ``simple`` as the module is manually installed.

The ``apply_extra`` script here requires
``unsquashfs`` to run, so that is built and installed.

.. note::

  Flatpak supports exporting icons, desktop files etc. from
  ``/app/extra/export/share/`` however when creating a Flatpak package
  for a software store like Flathub or when composing the metadata with
  appstreamcli, desktop files and icons must not be placed in
  ``/app/extra/export/`` and instead should go to the proper locations
  documented in :doc:`conventions`.

The icon, desktop file and Metainfo must be added as file sources. The
``path`` indicates that they reside in the same directory relative to the
manifest. The first three lines of the ``build-commands`` installs them
so that ``appstream-compose`` can compose the metadata.

The last line creates an empty symlink from ``${FLATPAK_DEST}/extra/``
to ``${FLATPAK_DEST}/bin/`` so that the executable is found in ``$PATH``
inside the sandbox and can be used in the top-level ``command: foo_binary``
property. Instead of a symlink this also often a script like:

.. code-block:: shell

  #!/bin/sh

>   exec /app/extra/subdir/foo_binary

Reviewer: ```suggestion
  exec /app/extra/subdir/foo_binary "$@"
```

Usually you always want to add `"$@"`.

direcory are moved to ``/app/extra`` and lastly ``squashfs-root`` and the
snap package itself is cleaned up to save space.

This is another example that can be used to extract Debian packages (``.deb``).

.. code-block:: yaml

  - name: foo
    buildsystem: simple
    build-commands:
      - install -Dm755 apply_extra ${FLATPAK_DEST}/bin/apply_extra
      - install -Dm644 -t ${FLATPAK_DEST}/share/metainfo/ ${FLATPAK_ID}.metainfo.xml
      - install -Dm644 -t ${FLATPAK_DEST}/share/applications/ ${FLATPAK_ID}.desktop
      - install -Dm644 -t ${FLATPAK_DEST}/share/icons/hicolor/scalable/apps/ ${FLATPAK_ID}.svg
      - ln -s ${FLATPAK_DEST}/extra/bin/foo_binary ${FLATPAK_DEST}/bin/foo_binary

    sources:
      - type: script
        dest-filename: apply_extra
        commands:
          - bsdtar --to-stdout -xf foo.deb data.* | bsdtar -xf -
          - rm foo.deb

Flatpak extensions can also use extra-data sources in a similar manner.
This is an example of an extension manifest utilising an extra-data
source.

.. code-block:: yaml

  app-id: com.foo.app.Plugins.audio
  runtime: com.foo.app
  sdk: org.freedesktop.Sdk//23.08
  build-extension: true
  separate-locales: false

>   appstream-compose: false

Reviewer: ```suggestion
```

we have established that this is no longer necessary.

        - install -Dm755 apply_extra -t $FLATPAK_DEST/bin/
      sources:
        - type: script
          dest-filename: apply_extra
          commands:
            - bsdtar -xf foo.qdz --strip-components=3
            - rm -f foo.qdz
        - type: extra-data
          filename: foo.qdz
          url: http://example.org/foo_v1.qdz
          sha256: 2027ed7e5935612a38c3f7b751b6bd23cfb5b06d1a1e5f34907bc021b09de225
          size: 523360406

For more information on Flatpak extensions please see :doc:`extension`.

Directory source
----------------

A directory source with ``type: dir`` is best suited for use in CI to
build the latest changes and avoid downloading multiple times. These
don't support any caching, so it will be rebuilt each time the
application is being built.

When submitting an application to software stores like Flathub, ``dir``
should be avoided altogether.

``path`` should be the path of the local directory relative to the
manifest root path, whoose contents will be copied during build.

.. code-block:: yaml

  modules:
    - name: foo
      buildsystem: simple
      build-commands:
        - ...
      sources:
        - type: dir
          path: icons

> Additonally there are `bzr`, `svn` and `data` sources supported. Please

Reviewer 2: Maybe mention some formats require specific tools are installed. Flathub doesn’t have bzr iirc.

Reviewer: I think it's not `data` but `inline`.

Author: Fixed
